### PR TITLE
Port material containers to datum components

### PR DESCRIPTION
**Pull request by Cyberboss**
Now anything can be a material container, instead of just some things.
Same code, just access it with GetComponent(/datum/component/material_container) instead of some materials var

### CODE REVIEW

File: code/datums/components/material_container.dm [hunk 1]
```diff
@@ -31,12 +43,44 @@
...
 
-/datum/material_container/Destroy()
-	owner = null
-	return ..()
+/datum/component/material_container/proc/OnExamine(mob/user)
+	for(var/I in materials)
+		var/datum/material/M = materials[I]
+		to_chat(user, "<span class='notice'>It has [amount(M.id)] units of [lowertext(M.name)] stored.</span>")
+
+/datum/component/material_container/proc/OnAttackBy(obj/item/I, mob/living/user)
```
Comment (Cyberboss) [inline code/datums/components/material_container.dm]: Majority of duplicated attackby code now goes here, reducing copypasta all around

**Comment by kevinz000** [comment]
and here i thought it was only for mobs.

**Comment by Tacolizard** [comment]
>When bae tells you she was tabs but she was actually spaces

**Comment by outbools** [comment]
Every object can be made from components according to ECS pattern! This will raise the level of abstraction and reduce copypasta, but 
how it gonna affect perfomance is a big question

**Comment by Jalleo** [comment]
@outbools it depends on many things. One of them the byond engine code.

File: code/game/atoms.dm [hunk 2]
```diff
@@ -290,6 +290,7 @@
 				to_chat(user, "[total_volume] units of various reagents")
 		else
 			to_chat(user, "Nothing.")
+	SendSignal(COMSIG_PARENT_EXAMINE, user)
```
Comment (AnturK) [inline code/game/atoms.dm]: This should have explicit order defined somewhere.

File: code/game/machinery/autolathe.dm [hunk 3]
```diff
@@ -135,42 +131,25 @@
...
 
-	busy = TRUE
-	var/inserted = materials.insert_item(O)
-	if(inserted)
-		if(istype(O, /obj/item/stack))
-			if (O.materials[MAT_METAL])
+/obj/machinery/mecha_part_fabricator/ComponentActivated(datum/component/C)
+	if(istype(C, /datum/component/material_container))
+		var/datum/component/material_container/M = C
+		if(!M.last_insert_success)
```
Comment (AnturK) [inline code/game/machinery/autolathe.dm]: Won't these be true if someone inserts a thing then examines without doing anything else.

File: code/game/atoms.dm (hunk 2 above)
Comment (Cyberboss) [inline code/game/atoms.dm]: Execution order for the components? or the examine?

File: code/game/machinery/autolathe.dm (hunk 3 above)
Comment (Cyberboss) [inline code/game/machinery/autolathe.dm]: You're right :thinking:

File: code/game/atoms.dm (hunk 2 above)
Comment (AnturK) [inline code/game/atoms.dm]: Either really. I was mostly thinking that examine order should be constant and controlled and afaik signal order would depend strictly on types of components here ?

File: code/game/machinery/autolathe.dm (hunk 3 above)
Comment (Cyberboss) [inline code/game/machinery/autolathe.dm]: Actually it's fine, ComponentActivated won't be called for examine because it doesn't return TRUE

Comment (AnturK) [inline code/game/machinery/autolathe.dm]: Still should probably be cleared somewhere, components shouldn't really rely on being alone.

File: code/game/atoms.dm (hunk 2 above)
Comment (Cyberboss) [inline code/game/atoms.dm]: Yeah, it would be determined by the order that components are added. I don't really have any ideas for how to control this...

**Comment by Cyberboss** [comment]
It needs a testmerge you fruit

**Comment by kevinz000** [comment]
finish this already :<

File: code/game/atoms.dm (hunk 2 above)
Comment (Cyberboss) [inline code/game/atoms.dm]: I suppose I could add a priority var to signals

**Comment by KorPhaeron** [comment]
@Cyberboss 

Tick tock

**Comment by Cyberboss** [comment]
Did some testing myself and all seems well

File: code/datums/components/README.md [hunk 4]
```diff
@@ -105,4 +105,8 @@ Stands have a lot of procs which mimic mob procs. Rather than inserting hooks fo
     * Called when a component recieves any signal and is enabled
     * Default implementation looks if the signal is registered and runs the appropriate proc
 
+<<<<<<< HEAD
+### See signals and their arguments in __DEFINES\components.dm
+=======
```
Comment (AnturK) [inline code/datums/components/README.md]: Ahem.

**Pull request merged**